File: 10 November 2022/Python/Jumping Numbers.py
#User function Template for python3

class Solution:
    def jumpingNums(self, X):
        if X<=9:
            return X
        diff = X
        ans = 0
        def solve(v):
            nonlocal ans,diff
            if v>X:
                return 
            if v<= X and v>ans:
                ans = v
                diff = X-v
            rem = v%10
            if rem>0:
                solve(v*10 + rem-1)
            if rem<9:
                solve(v*10 + rem+1)
            return 
        for i in range(1,10):
            solve(i)
        return ans

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        X=int(input())
        
        ob = Solution()
        print(ob.jumpingNums(X))
# } Driver Code Ends

File: November 2022/10 November 2022/Java/Jumping Numbers.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {
            long X = Long.parseLong(read.readLine());

            Solution ob = new Solution();
            System.out.println(ob.jumpingNums(X));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution {
    static long jumpingNums(long X) {
        // code here
        ArrayList<Long> res = new ArrayList<>();
        for (long i = 1; i <= 9 && i <= X; i++) {
            bfs(X, i,res);
        }
       return Collections.max(res);
    }
    
    public static void bfs(long x, long num, ArrayList<Long> res)
    {
        // Create a queue and enqueue 'i' to it
        Queue<Long> q = new LinkedList<Long>();
        q.add(num);

        
        // Do BFS starting from i
        while (!q.isEmpty()) {
            num = q.peek();
            q.poll();
            if (num <= x) {
                res.add(num);
                long last_digit = num % 10;

                // If last digit is 0, append next digit only
                if (last_digit == 0) {
                    q.add((num * 10) + (last_digit + 1));
                }

                // If last digit is 9, append previous digit only
                else if (last_digit == 9) {
                    q.add((num * 10) + (last_digit - 1));
                }

                // If last digit is neither 0 nor 9, append both
                // previous and next digits
                else {
                    q.add((num * 10) + (last_digit - 1));
                    q.add((num * 10) + (last_digit + 1));
                }
                
                
            }
        }
    }
};